marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs

```diff
@@ -0,0 +1,564 @@
+import Hasql.Session (Session, sql, statement)
+import Hasql.Statement (refineResult)
+import Hasql.TH (maybeStatement, resultlessStatement, vectorStatement)
+import Language.Marlowe.Runtime.ChainSync.Database (CardanoBlock, CommitBlocks (..), CommitGenesisBlock (..),
+                                                    CommitRollback (..), DatabaseQueries (DatabaseQueries),
+                                                    GetGenesisBlock (..), GetHeaderAtPoint (..),
+                                                    GetIntersectionPoints (..))
+import Language.Marlowe.Runtime.ChainSync.Genesis (GenesisBlock (..), GenesisTx (..))
+import Ouroboros.Network.Point (WithOrigin (..))
+
+-- | PostgreSQL implementation for the chain sync database queries.
+databaseQueries :: DatabaseQueries Session
+databaseQueries = DatabaseQueries
+ commitRollback
+ commitBlocks
+ commitGenesisBlock
+ getHeaderAtPoint
+ getIntersectionPoints
+ getGenesisBlock
+
+-- GetGenesisBlock
+
+getGenesisBlock :: GetGenesisBlock Session
+getGenesisBlock = GetGenesisBlock $ statement () $ refineResult (decodeResults . V.toList)
+  [vectorStatement|
+    SELECT block.id :: bytea
+         , tx.id :: bytea
+         , txOut.lovelace :: bigint
+         , txOut.address :: bytea
+    FROM   chain.block AS block
+    JOIN   chain.tx    AS tx    ON tx.blockId = block.id
+                               AND tx.slotNo = block.slotNo
+    JOIN   chain.txOut AS txOut ON txOut.txId   = tx.id
+    WHERE  block.slotNo = -1
+  |]
+  where
+    decodeResults :: [(ByteString, ByteString, Int64, ByteString)] -> Either Text (Maybe GenesisBlock)
+    decodeResults [] = Right Nothing
+    decodeResults rows@((hash, _, _, _) : _) =
+      Just . GenesisBlock (HeaderHash $ toShort hash) . Set.fromList <$> traverse decodeGenesisTx rows
+
+    decodeGenesisTx :: (ByteString, ByteString, Int64, ByteString) -> Either Text GenesisTx
+    decodeGenesisTx (_, txId, lovelace, address) = GenesisTx
+      <$> decodeTxId txId
+      <*> pure (fromIntegral lovelace)
+      <*> decodeAddressAny address
+
+-- GetHeaderAtPoint
+
+getHeaderAtPoint :: GetHeaderAtPoint Session
+getHeaderAtPoint = GetHeaderAtPoint \case
+  ChainPointAtGenesis -> pure Origin
+  point@(ChainPoint slotNo hash) ->
+    statement (slotNoToParam slotNo, headerHashToParam hash) $ refineResult decodeResults
+      [maybeStatement|
+        SELECT block.blockNo :: bigint
+          FROM chain.block AS block
+        WHERE block.slotNo = $1 :: bigint
+          AND block.id     = $2 :: bytea
+      |]
+    where
+      decodeResults :: Maybe Int64 -> Either Text (WithOrigin BlockHeader)
+      decodeResults Nothing        = Left $ "No block found at " <> T.pack (show point)
+      decodeResults (Just blockNo) = Right $ At $ BlockHeader slotNo hash $ BlockNo $ fromIntegral blockNo
+
+decodeTxId :: ByteString -> Either Text TxId
+decodeTxId txId = case deserialiseFromRawBytes AsTxId txId of
+  Nothing    -> Left "Invalid TxId"
```

Comment:
Should we include the bytes that couldn't be decoded in the error message itself?

Reply:
We could